main.go

```diff
@@ -3,126 +3,160 @@ package main
 	flag.StringVar(&cloudURI, "cloud-uri", "", "Cloud URI that edge tunneling service is running on")
 	flag.StringVar(&apiKey, "api-key", "", "API Key or JWT Token that needs to connect to edge tunneling service")
 	flag.StringVar(&applicationID, "application-id", "", "If JWT is used instead of api-key, application-id is needed")
+	flag.IntVar(&pingDuration, "ping-duration", 5, "Ping duration in seconds, 0 will disable pings to the tunnel")
 	flag.Parse()
 
+	customFormatter := new(logrus.TextFormatter)
+	customFormatter.FullTimestamp = true
+	logrus.SetFormatter(customFormatter)
+	logrus.SetLevel(logrus.DebugLevel)
+
 	if cloudURI == "" {
-		fmt.Printf("cloud-uri must be provided\n")
+		logrus.Errorf("cloud-uri must be provided\n")
 
 		os.Exit(1)
 	}
 
 	if apiKey == "" {
-		fmt.Printf("api-key must be provided\n")
+		logrus.Errorf("api-key must be provided\n")
 
 		os.Exit(1)
 	}
 
 	// setup tcp server
 	listener, err := net.Listen("tcp", listenURI)
 	if err != nil {
-		fmt.Printf("Error listening: %s\n", err.Error())
+		logrus.Errorf("Error listening: %s\n", err.Error())
 
 		os.Exit(1)
 	}
 	defer listener.Close()
 
-	fmt.Printf("Listening on: %s\n", listenURI)
+	if pingDuration > 0 {
+		logrus.Infof("Send pings enabled with an interval %v seconds\n", pingDuration)
+	} else {
+		logrus.Infof("Send pings disabled\n")
+	}
+
+	logrus.Infof("Listening on: %s\n", listenURI)
 
 	ctx, cancel := context.WithCancel(context.Background())
 
-	go func() {
-		defer cancel()
+	defer cancel()
 
-		for {
-			tcpConn, err := listener.Accept()
-			if err != nil {
-				fmt.Printf("Error accepting: %s\n", err.Error())
+	for {
+		tcpConn, err := listener.Accept()
+		if err != nil {
+			logrus.Errorf("Error accepting: %s\n", err.Error())
 
-				return
-			}
+			return
+		}
 
-			fmt.Printf("Received message from %s to %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
+		logrus.Debugf("(Connection from %s to %s) Starting\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr())
 
-			wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
-				"Authorization": []string{
-					"Bearer " + apiKey,
-				},
-				"X-Application-ID": []string{applicationID},
-			})
+		wsConn, _, err := websocket.DefaultDialer.Dial(cloudURI, http.Header{
+			"Authorization": []string{
+				"Bearer " + apiKey,
+			},
+			"X-Application-ID": []string{applicationID},
+		})
 
-			if err != nil {
-				fmt.Printf("Error dial into edge tunneling service: %s\n, cloudURI - %s\n", err.Error(), cloudURI)
+		if err != nil {
+			logrus.Errorf("(Connection from %s to %s) Error dial into edge tunneling service: %s\n, cloudURI - %s\n", tcpConn.RemoteAddr(), tcpConn.LocalAddr(), err.Error(), cloudURI)
 
-				return
-			}
+			return
+		}
 
-			ctx, cancel := context.WithCancel(ctx)
+		ctx, cancel := context.WithCancel(ctx)
 
-			go func() {
-				for {
+		go func() {
+			chanBytes := make(chan []byte, 4096)
+			go func () {
+				for{
```

Comment:
should we add a listener here as well to monitor `ctx.Done()`? As it might be cancelled by other go routines.

Reply:
Keeping track of the error sequence in the logs should be fine. Also saves us another channel.